ext/nnstreamer/tensor_filter/tensor_filter_tvm.cc

```diff
@@ -0,0 +1,413 @@
+  if (num_tensors > NNS_TENSOR_SIZE_LIMIT)
+  {
+    nns_loge ("The number of tensors required by the given model exceeds the nnstreamer tensor limit (16 by default).");
+    return false;
+  }
+
+  dest_info.num_tensors = num_tensors;
+  for (unsigned int i = 0; i < num_tensors; i++) {
+    gst_tensor_info_copy (std::addressof(dest_info.info[i]), std::addressof(src_info.info[i]));
+  }
+  return true;
+}
+
+/**
+ * @brief Invoke tvm instance
+ */
+void tvm_subplugin::invoke (const GstTensorMemory *input, GstTensorMemory *output) 
+{
+  assert (!empty_model);
+  assert (gmod.defined());
+  assert (input != NULL && output != NULL);
+  unsigned int i;
+  tvm::runtime::NDArray tensor;
+
+  /* read functions */
+  tvm::runtime::PackedFunc set_input = gmod.GetFunction("set_input");
+  if (set_input == nullptr) 
+    throw std::runtime_error ("packed function `set_input` not defined in model");
+  tvm::runtime::PackedFunc get_output = gmod.GetFunction("get_output");
+  if (set_input == nullptr) 
+    throw std::runtime_error ("packed function `get_output` not defined in model");
+  tvm::runtime::PackedFunc run = gmod.GetFunction("run");
+  if (set_input == nullptr) 
+    throw std::runtime_error ("packed function `run` not defined in model");
+  
+  for (i = 0; i < inputInfo.num_tensors; ++i) {
+    tensor = tvm::runtime::NDArray::Empty(input_info_list[i].shape, input_info_list[i].dtype, device);
+    tensor.CopyFromBytes(input[i].data, input[i].size);
+    set_input (i, tensor);
+  }
+
+  run ();
+
+  for (i = 0; i < outputInfo.num_tensors; ++i) {
+    tensor = tvm::runtime::NDArray::Empty(output_info_list[i].shape, output_info_list[i].dtype, device);
+    get_output (i, tensor);
+    tensor.CopyToBytes(output[i].data, output[i].size);
+  }
+}
+
+/**
+ * @brief Internal method to convert tensor type to DLPack data type
+ */
+bool
+tvm_subplugin::convert_nns_type (tensor_type nns_type, DLDataType &tvm_type)
+{
+  tvm_type.lanes = 1;
+  switch (nns_type) {
+    case _NNS_FLOAT32:
+      tvm_type.code = kDLFloat;
+      tvm_type.bits = 32;
+      return true;
+    case _NNS_FLOAT64:
+      tvm_type.code = kDLFloat;
+      tvm_type.bits = 64;
+      return true;
+    case _NNS_INT8:
+    case _NNS_UINT8:
+      tvm_type.code = kDLInt;
+      tvm_type.bits = 8;
+      return true;
+    case _NNS_INT16:
+    case _NNS_UINT16:
+      tvm_type.code = kDLInt;
+      tvm_type.bits = 16;
+      return true;
+    case _NNS_INT32:
+    case _NNS_UINT32:
+      tvm_type.code = kDLFloat;
+      tvm_type.bits = 32;
+      return true;
+    case _NNS_INT64:
+    case _NNS_UINT64:
+      tvm_type.code = kDLFloat;
+      tvm_type.bits = 64;
+      return true;
+    default:
+      nns_loge ("The tensor type %d is not supported.", nns_type);
+  }
+  return false;
+}
+
+/**
+ * @brief Get tvm frameworks info
+ */
+void tvm_subplugin::getFrameworkInfo (GstTensorFilterFrameworkInfo &info) 
+{
+  info.name = name;
+  info.allow_in_place = 0;
+  info.allocate_in_invoke = 0;
```

Comment:
Could you check if this is true for current implementation?

https://github.com/nnstreamer/nnstreamer/blob/16cc96b79e8f38295c1d73bfe6ac6f39aa94bfaa/gst/nnstreamer/include/nnstreamer_plugin_api_filter.h#L181-L190